src/zero_copy/pratt.rs

```diff
@@ -0,0 +1,363 @@
+    type Strength: Copy + Ord;
+
+    /// Get the parsing precedence of this operator.
+    ///
+    /// If an expression has an operator on both sides,
+    /// the one with the greatest strength will
+    /// be built first.
+    ///
+    /// For example, given `x + y * z` where `*` has a greater strength
+    /// than `+` (as usual), the `y` will be combined with the `z` first.
+    /// Next, the combination `(y * z)` will be combined with `x`,
+    /// resulting in `(x + (y * z))`.
+    ///
+    /// If both sides have operators with the same strength,
+    /// then the associativity will determine which side
+    /// will be combined first.
+    ///
+    /// For example, given `x + y + z`;
+    /// left-associativity will result in `((x + y) + z)`,
+    /// while right-associativity will result in `(x + (y + z))`,
+    fn precedence(&self) -> InfixPrecedence<Self::Strength>;
+
+    /// Build an expression for this operator given two arguments.
+    fn build_expression(self, left: Expr, right: Expr) -> Expr;
+}
+
+/// See [`Parser::pratt`].
+#[derive(Copy, Clone)]
+pub struct Pratt<Extra, AtomParser, Expr, OpParser, Op> {
+    pub(crate) parser_op: OpParser,
+    pub(crate) parser_atom: AtomParser,
+    pub(crate) phantom: PhantomData<(Extra, Expr, Op)>,
+}
+
+impl<E, AtomParser, Expr, OpParser, Op> Pratt<E, AtomParser, Expr, OpParser, Op> {
+    fn pratt_parse<'a, M, I>(
+        &self,
+        inp: &mut InputRef<'a, '_, I, E>,
+        min_strength: Option<Strength<Op::Strength>>,
+    ) -> PResult<M, Expr, E::Error>
+    where
+        I: Input + ?Sized,
+        E: ParserExtra<'a, I>,
+        AtomParser: Parser<'a, I, Expr, E>,
+        OpParser: Parser<'a, I, Op, E>,
+        Op: InfixOperator<Expr>,
+        M: Mode,
+    {
+        let mut left = self.parser_atom.go::<M>(inp)?;
+        loop {
+            let pre_op = inp.save();
+            let (op, prec) = match self.parser_op.go::<Emit>(inp) {
+                Ok(op) => {
+                    let prec = op.precedence();
+                    if prec.strength_left().is_lt(&min_strength) {
+                        inp.rewind(pre_op);
+                        return Ok(left);
+                    }
+                    (op, prec)
+                }
+                Err(_) => return Ok(left),
+            };
+
+            let right = self.pratt_parse::<M, _>(inp, Some(prec.strength_right()))?;
+            left = M::combine(left, right, |l: Expr, r: Expr| op.build_expression(l, r));
+        }
+    }
+}
+
+impl<'a, I, E, Expr, AtomParser, OpParser, Op> Parser<'a, I, Expr, E>
+    for Pratt<E, AtomParser, Expr, OpParser, Op>
+where
+    I: Input + ?Sized,
+    E: ParserExtra<'a, I>,
+    AtomParser: Parser<'a, I, Expr, E>,
+    OpParser: Parser<'a, I, Op, E>,
+    Op: InfixOperator<Expr>,
+{
+    fn go<M: Mode>(&self, inp: &mut InputRef<'a, '_, I, E>) -> PResult<M, Expr, E::Error>
+    where
+        Self: Sized,
+    {
+        self.pratt_parse::<M, _>(inp, None)
```

Comment:
Would it make sense to just inline `pratt_parse` here? I can't think of a reason to call it outside of this context anyway given that it already needs all of the 'mucky internals' associated with parsing (like `InputRef`).

Reply:
I may be missing something obvious, but due to it's recursive nature requiring the `None` on the first call, and `Some(prec.strength_right())` on following calls I don't see how that's possible to inline the body into `Parser::go`. Unless you literally just mean adding the `#[inline]` tag, in which case that's not a problem :D

Reply:
Ah, no problem, I'm being a fool